// @nativescript/schematics/src/ng-new/shared/_sample-files/barcelona/player-detail/player-detail.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var router_1 = require("@angular/router");
var player_service_1 = require("../player.service");
var PlayerDetailComponent = /** @class */ (function () {
    function PlayerDetailComponent(playerService, route) {
        this.playerService = playerService;
        this.route = route;
    }
    PlayerDetailComponent.prototype.ngOnInit = function () {
        var id = +this.route.snapshot.params['id'];
        this.player = this.playerService.getPlayer(id);
    };
    PlayerDetailComponent = __decorate([
        core_1.Component({
            selector: '<%= prefix %>-details',
            templateUrl: './player-detail.component.html',
        }),
        __metadata("design:paramtypes", [player_service_1.PlayerService,
            router_1.ActivatedRoute])
    ], PlayerDetailComponent);
    return PlayerDetailComponent;
}());
exports.PlayerDetailComponent = PlayerDetailComponent;
